// @nativescript/angular/esm2022/lib/cdk/frame-page/frame-page.component.mjs
import { Component, ElementRef, Input, Renderer2 } from '@angular/core';
import { Frame, Page } from '@nativescript/core';
import { frameMeta, registerElement } from '../../element-registry';
import { PageService } from './page.service';
import * as i0 from "@angular/core";
export function customFrameComponentFactory(v) {
    return v.element.nativeElement;
}
export function customPageFactoryFromFrame(v) {
    return v.page;
}
registerElement('FramePage', () => Frame, {
    insertChild: (parent, child) => {
        if (child instanceof Page) {
            frameMeta.insertChild(parent, child);
            return;
        }
        parent['__ng_page__'].content = child;
    },
});
class FramePageComponent {
    constructor(element, renderer) {
        this.element = element;
        this.actionBarHidden = false;
        this.page = renderer.createElement('Page');
        element.nativeElement['__ng_page__'] = this.page;
        renderer.appendChild(element.nativeElement, this.page);
    }
    ngOnChanges(changes) {
        if (changes.actionBarHidden && changes.actionBarHidden.previousValue !== changes.actionBarHidden.currentValue) {
            if (this.page) {
                this.page.actionBarHidden = !!this.actionBarHidden;
            }
        }
    }
    ngAfterViewInit() {
        if (this.page) {
            this.page.actionBarHidden = !!this.actionBarHidden;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: FramePageComponent, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.0.1", type: FramePageComponent, selector: "FramePage", inputs: { actionBarHidden: "actionBarHidden" }, providers: [
            {
                provide: Frame,
                useFactory: customFrameComponentFactory,
                deps: [FramePageComponent],
            },
            {
                provide: Page,
                useFactory: customPageFactoryFromFrame,
                deps: [FramePageComponent],
            },
            PageService,
        ], usesOnChanges: true, ngImport: i0, template: `<ng-content></ng-content>`, isInline: true }); }
}
export { FramePageComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: FramePageComponent, decorators: [{
            type: Component,
            args: [{
                    // eslint-disable-next-line @angular-eslint/component-selector
                    selector: 'FramePage',
                    template: `<ng-content></ng-content>`,
                    providers: [
                        {
                            provide: Frame,
                            useFactory: customFrameComponentFactory,
                            deps: [FramePageComponent],
                        },
                        {
                            provide: Page,
                            useFactory: customPageFactoryFromFrame,
                            deps: [FramePageComponent],
                        },
                        PageService,
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, propDecorators: { actionBarHidden: [{
                type: Input
            }] } });
