// @nativescript/angular/esm2022/lib/cdk/frame-page/frame.directive.mjs
import { Directive, ElementRef } from '@angular/core';
import { Frame } from '@nativescript/core';
import * as i0 from "@angular/core";
export function customFrameDirectiveFactory(v) {
    return v.element.nativeElement;
}
class FrameDirective {
    constructor(element) {
        this.element = element;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: FrameDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.1", type: FrameDirective, selector: "Frame", providers: [
            {
                provide: Frame,
                useFactory: customFrameDirectiveFactory,
                deps: [FrameDirective],
            },
        ], ngImport: i0 }); }
}
export { FrameDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: FrameDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'Frame',
                    providers: [
                        {
                            provide: Frame,
                            useFactory: customFrameDirectiveFactory,
                            deps: [FrameDirective],
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; } });
