// @nativescript/angular/esm2022/lib/views/invisible-nodes.mjs
import { View } from '@nativescript/core';
const getClassName = (instance) => instance.constructor.name;
export class InvisibleNode extends View {
    constructor(name = '') {
        super();
        this.name = name;
        this.nodeType = 1;
        this.nodeName = getClassName(this);
    }
    toString() {
        return `${this.nodeName}(${this.id})-${this.name}`;
    }
}
class CommentNode extends InvisibleNode {
    static { this.id = 0; }
    constructor(value) {
        super(value);
        this.meta = {
            skipAddToDom: true,
        };
        this.id = CommentNode.id.toString();
        CommentNode.id += 1;
    }
}
export { CommentNode };
class TextNode extends InvisibleNode {
    static { this.textChangeEvent = 'textChange'; }
    static { this.id = 0; }
    get text() {
        return this._text;
    }
    set text(t) {
        this._text = t;
        this.notify({ eventName: TextNode.textChangeEvent, object: this, value: t });
    }
    constructor(value) {
        super(value);
        this._text = '';
        this.callbackMap = new Map();
        this._text = value;
        this.meta = {
            skipAddToDom: true,
        };
        this.id = TextNode.id.toString();
        TextNode.id += 1;
    }
    registerTextChange(callback, id) {
        const cb = (evt) => callback(evt.value);
        const cbArr = this.callbackMap.get(id) || [];
        cbArr.push(cb);
        this.callbackMap.set(id, cbArr);
        this.on('textChange', cb);
    }
    unregisterTextChange(id) {
        const cbArr = this.callbackMap.get(id) || [];
        cbArr.forEach((cb) => this.off('textChange', cb));
        this.callbackMap.delete(id);
    }
}
export { TextNode };
