// @nativescript/angular/esm2022/lib/rootcomposite.module.mjs
import { NgModule } from '@angular/core';
import { Application, ContentView, RootLayout } from '@nativescript/core';
import { APP_ROOT_VIEW, DISABLE_ROOT_VIEW_HANDLING } from './tokens';
import * as i0 from "@angular/core";
export class RootViewProxy extends ContentView {
    constructor(parentRootLayout) {
        super();
        this.parentRootLayout = parentRootLayout;
    }
    _addView(view, atIndex) {
        super._addView(view, atIndex);
        if (this.parentRootLayout.getChildIndex(this) < 0) {
            this.parentRootLayout.insertChild(this, 0);
        }
    }
    _removeView(view) {
        super._removeView(view);
        this.parentRootLayout.removeChild(this);
    }
}
/**
 * This generates a RootLayout and returns a RootViewProxy.
 * Setting RootViewProxy.content will add the view to the bottom of the RootLayout
 * Setting RootViewProxy.content = null will remove the view from the RootLayout
 * @returns RootViewProxy that will insert content into the start of the RootLayout
 */
export function generateRootLayoutAndProxy() {
    let rootView = Application.getRootView();
    if (!rootView || !(rootView instanceof RootLayout)) {
        rootView = new RootLayout();
        Application.resetRootView({ create: () => rootView });
    }
    const viewProxy = new RootViewProxy(rootView);
    // (rootView as RootLayout).insertChild(viewProxy, 0);
    return viewProxy;
}
class RootCompositeModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: RootCompositeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.1", ngImport: i0, type: RootCompositeModule }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: RootCompositeModule, providers: [
            { provide: DISABLE_ROOT_VIEW_HANDLING, useValue: true },
            { provide: APP_ROOT_VIEW, useFactory: generateRootLayoutAndProxy },
        ] }); }
}
export { RootCompositeModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: RootCompositeModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [
                        { provide: DISABLE_ROOT_VIEW_HANDLING, useValue: true },
                        { provide: APP_ROOT_VIEW, useFactory: generateRootLayoutAndProxy },
                    ],
                }]
        }] });
