// @nativescript/angular/esm2022/lib/trace.mjs
import { Trace } from '@nativescript/core';
class NativeScriptDebug {
    static { this.animationsTraceCategory = 'ns-animations'; }
    static { this.rendererTraceCategory = 'ns-renderer'; }
    static { this.viewUtilCategory = 'ns-view-util'; }
    static { this.routerTraceCategory = 'ns-router'; }
    static { this.routeReuseStrategyTraceCategory = 'ns-route-reuse-strategy'; }
    static { this.listViewTraceCategory = 'ns-list-view'; }
    static { this.bootstrapCategory = 'bootstrap'; }
    // TODO: migrate all usage to this - avoids extraneous method executions
    static { this.enabled = Trace.isEnabled(); }
    static isLogEnabled() {
        return Trace.isEnabled();
    }
    static animationsLog(message) {
        Trace.write(message, NativeScriptDebug.animationsTraceCategory);
    }
    static rendererLog(msg) {
        Trace.write(msg, NativeScriptDebug.rendererTraceCategory);
    }
    static rendererError(message) {
        Trace.write(message, NativeScriptDebug.rendererTraceCategory, Trace.messageType.error);
    }
    static viewUtilLog(msg) {
        Trace.write(msg, NativeScriptDebug.viewUtilCategory);
    }
    static routerLog(message) {
        Trace.write(message, NativeScriptDebug.routerTraceCategory);
    }
    static routerError(message) {
        Trace.write(message, NativeScriptDebug.routerTraceCategory, Trace.messageType.error);
    }
    static routeReuseStrategyLog(message) {
        Trace.write(message, NativeScriptDebug.routeReuseStrategyTraceCategory);
    }
    static styleError(message) {
        Trace.write(message, Trace.categories.Style, Trace.messageType.error);
    }
    static listViewLog(message) {
        Trace.write(message, NativeScriptDebug.listViewTraceCategory);
    }
    static listViewError(message) {
        Trace.write(message, NativeScriptDebug.listViewTraceCategory, Trace.messageType.error);
    }
    static bootstrapLog(message) {
        Trace.write(message, NativeScriptDebug.bootstrapCategory);
    }
    static bootstrapLogError(message) {
        Trace.write(message, NativeScriptDebug.bootstrapCategory, Trace.messageType.error);
    }
}
export { NativeScriptDebug };
