// @nativescript/angular/esm2022/lib/cdk/platform-filters/vision-filter.component.mjs
/* eslint-disable @angular-eslint/component-selector */
import { Component, Inject } from '@angular/core';
import { platformNames } from '@nativescript/core';
import { DEVICE } from '../../tokens';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
class VisionOSFilterComponent {
    constructor(device) {
        // Note: casting any to be backwards compatible
        // Available in 8.6+ only
        this.show = device.os === platformNames.visionos;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: VisionOSFilterComponent, deps: [{ token: DEVICE }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.0.1", type: VisionOSFilterComponent, selector: "visionos", ngImport: i0, template: `<ng-content *ngIf="show"></ng-content>`, isInline: true, dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }] }); }
}
export { VisionOSFilterComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: VisionOSFilterComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'visionos',
                    template: `<ng-content *ngIf="show"></ng-content>`,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DEVICE]
                }] }]; } });
